scripts/firedrake-run-split-tests

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
+
+# Script for running a pytest test suite in parallel across multiple jobs.
+#
+# The test suite is partitioned between the jobs and each job outputs its own log file.
+#
+# Usage:
+#
+#     firedrake-run-split-tests <nprocs> <njobs> <pytest_args...>
+#
+#   where:
+#     * <nprocs> is the number of ranks used in each test
+#     * <njobs> is the number of different jobs
+#     * <pytest_args...> are additional arguments that are passed to pytest
+#
+# Example:
+#
+#   To give an example,
+#
+#     firedrake-run-split-tests 3 4 tests/unit --verbose
+#
+#   will run all of the parallel[3] tests inside tests/unit verbosely
+#   and split between 4 different jobs.
+#
+# Requires:
+#
+#   * pytest
+#   * pytest-split
+#   * pytest-mpi (https://github.com/firedrakeproject/pytest-mpi)
+#   * GNU parallel
+
+num_procs=$1
+num_jobs=$2
+extra_args=${@:3}
+
+log_file_prefix="pytest_nprocs${num_procs}_job"
+
+if [ $num_procs = 1 ]; then
+    pytest_exec="python -m pytest"
+    marker_spec="\"parallel[1] or not parallel\""
+else
+    pytest_exec="mpiexec -n ${num_procs} python -m pytest"
+    marker_spec="parallel[${num_procs}]"
+fi
+
+pytest_cmd="${pytest_exec} -v \
+            --splits ${num_jobs} --group {#} \
+            -m ${marker_spec} ${extra_args}"
+
+# This incantation:
+# * Runs pytest under GNU parallel using the right number of jobs
+# * Uses tee to pipe stdout+stderr to both stdout and a log file
+# * Writes pytest's exit code to a file called jobN.errcode (for later inspection)
+insn="parallel --line-buffer --tag \
+      ${pytest_cmd} |& tee ${log_file_prefix}{#}.log; \
+      echo \${PIPESTATUS[0]} > job{#}.errcode \
+      ::: $(seq ${num_jobs})"
+
+echo Running: ${insn}
```

Comment:
If you use `set -x` and `set +x` either side of running then you won't have to create the `insn` variable just to be able to print it.
I do like still keeping the `pytest_cmd` as a separate variable though, it makes the `parallel` command clearer.

Reply:
Can you show me an example of what you mean?

Reply:
From `man bash`:
`-x        Print commands and their arguments as they are executed.`

It literally just does what you are doing here, then `set +x` unsets the options after. You don't have to make a use-once variable, which also means you don't need to worry about escaping things in the insn string.